inferno-lsp/src/Inferno/LSP/ParseInfer.hs

```diff
@@ -508,77 +508,55 @@ parseAndInfer prelude idents txt validateInput = do
-                        Nothing -> return $ Left [errorDiagnosticInfer 0 0 0 2 $ renderDoc $ vsep ["Could not find the input that caused the error"]]
-                        Just (s, e) -> return $ Left [errorDiagnosticInfer (unPos $ sourceLine s) (unPos $ sourceColumn s) (unPos $ sourceLine e) (unPos $ sourceColumn e) err]
-                _ -> do
-                  -- liftIO $ debugM "reactor.handle" $ "Finished inferring"
-                  let final = insertCommentsIntoExpr comments pinnedAST'
-                  return $
-                    Right
-                      ( fmap (const ()) final,
-                        tcSch,
-                        Map.foldrWithKey (\k v xs -> (mkHover prelude currentClasses k v) : xs) mempty tyMap
-                      )
+              let signature = collectArrs typSig
+              -- Check that script isn't itself a function
+              case checkScriptIsNotAFunction signature idents of
+                Left errors -> return $ Left errors
+                Right () ->
+                  -- Validate input types
+                  case checkScriptInputTypes signature pinnedAST' of
+                    Left errors -> return $ Left errors
+                    Right () -> do
+                      let final = insertCommentsIntoExpr comments pinnedAST'
+                      return $
+                        Right
+                          ( fmap (const ()) final,
+                            tcSch,
+                            Map.foldrWithKey (\k v xs -> (mkHover prelude currentClasses k v) : xs) mempty tyMap
+                          )
   where
-    -- Example: The type signature has the format for 3 inputs:
-    -- TArr
-    --   Input1
-    --   TArr
-    --     Input2
-    --     TArr
-    --       Input3
-    --       Output
-    --
-    -- Therefore we need to check the type of the first value in each TArr to see if it is a valid input type
-    -- The final output value will be ignored
-    validateInputs :: InfernoType -> [Either Text ()]
-    validateInputs = \case
-      -- Ignore validation if the final output is an array
-      typSig | isOutputAnArray typSig -> [Right ()]
-      -- Head is input, tail *could* be other inputs
-      TArr h t | isTArr t -> validateInput h : validateInputs t
-      -- Tail of this would be the output
-      TArr h _t | not (isTArr h) -> [validateInput h]
-      -- Top level is not an arr, assume it is correct to prevent errors
-      _ -> [Right ()]
+    checkScriptIsNotAFunction signature parameters =
+      -- A function with N parameters should have a signature a_1 -> a_2 -> ... -> a_{N+1}
+      if length signature > (length parameters + 1)
+        then Left [errorDiagnosticInfer 0 0 0 2 $ renderDoc $ vsep ["This script evaluates to a function. Did you mean to add input parameters instead?"]]
+        else Right ()
 
-    isTArr :: InfernoType -> Bool
-    isTArr = \case
-      TArr _ _ -> True
-      _ -> False
```

Comment:
@LTibbetts I simplified these functions by using the inbuilt function `collectArrs` that converts an `InfernoType` potentially containing `TArr`s to a `[InfernoType]`. I couldn't think of a reason why input type validation should be ignored if the final output is an array, so I removed that check. (In fact, don't we disallow array outputs in OnPing anyway?)